### ISSUE TITLE
DXC function overloading doesn't behave correctly when the function argument type is a structure

### ISSUE THREAD

**Issue by jirzynek** [microsoft/DirectXShaderCompiler#1694]
When compiling this .hlsl code:
```
// dxc.exe out.txt -nologo -O3 -Zi -Qstrip_debug -Tps_6_3 -Etest_ps -Fo"out.bin"

struct TYPE_A
{
    float3 AAA;
};

struct TYPE_B
{
    float3 BBB;
};

void foo(TYPE_A psin)
{
}

void foo(TYPE_B psin)
{
}

float4 test_ps() : SV_Target0
{
    TYPE_A t;
    foo(t);
 
    return float4(0,0,0,0);
}
```
DXC will fail with the following error: 
```
out.txt:26:5: error: call to 'foo' is ambiguous
    foo(t);
    ^~~
out.txt:14:6: note: candidate function
void foo(const TYPE_A psin)
     ^
out.txt:18:6: note: candidate function
void foo(const TYPE_B psin)
     ^
```
however adding an additional field to TYPE_A structure fixes the problem, for example this will compile fine:
```
struct TYPE_A
{
    float3 AAA;
    float3 YYY;
};

struct TYPE_B
{
    float3 BBB;
};

void foo(TYPE_A psin)
{
}

void foo(TYPE_B psin)
{
}

float4 test_ps() : SV_Target0
{
    TYPE_A t;
    foo(t);
 
    return float4(0,0,0,0);
}
```

Should DXC follow C/C++ convention and treat TYPE_A and TYPE_B as a separate types or is it the expected behavior?

**Comment by MrTrillian** [comment]
Note that the behavior is consistent with FXC:

```
"C:\Program Files (x86)\Windows Kits\10\bin\10.0.17763.0\x64\fxc.exe" /E TestPS /T ps_5_0 overloading.hlsl
Microsoft (R) Direct3D Shader Compiler 10.1 (using C:\Program Files (x86)\Windows Kits\10\bin\10.0.17763.0\x64\D3DCOMPILER_47.dll)
Copyright (C) 2013 Microsoft. All rights reserved.

C:\Users\trlabell\Desktop\hlsl\overloading.hlsl(22,5-10): error X3067: 'foo': ambiguous function call

compilation failed; no code produced
```

Note that you also get that error with `struct TYPE_A { float A; struct { float2 AA; } S; }`, for example. Basically the HLSL type system is based on structural rather than nominal equivalence of structs.

**Comment by jirzynek** [comment]
@MrTrillian  indeed, it's consistent, but I was hoping that new DXC can improve on some things :) I know this is HLSL feature but when writing a function that accepts type A, I'd expect to accept it only type A and nothing else. Maybe there is a possibility of adding an optional warning to DXC (cannot convert argument from TYPE_B to TYPE_A)?

This for example compiles fine on FXC, but does it make sense?
```
struct TYPE_POSITION
{
    float3 position : POS;
};

struct TYPE_NORMAL
{
    float3 normal : NORMAL;
};

void calculate_position(TYPE_POSITION psin)
{
}

float4 test_ps() : SV_Target0
{
    TYPE_POSITION ta;
    TYPE_NORMAL tb;
    calculate_position(ta);
    calculate_position(tb);
 
    return float4(0,0,0,0);
}

```

**Comment by kayru** [comment]
I think at least for us, C++-like behavior would be the most desirable (least surprising). We would love an ability to opt out of legacy FXC behavior here.

**Comment by TheRealMJP** [comment]
I agree with Yuriy on this one: I would strongly prefer that DXC were more consistent with C++ with overloading. An option when compiling would be fine, we would always enable it.

**Comment by d11mshepcar** [comment]
+1 for an opt out on this!

**Comment by pow2clk** [comment]
This was fixed by #3891 as part of the HLSL 2021 language revision!
